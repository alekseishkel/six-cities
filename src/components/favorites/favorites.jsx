import React from 'react';
import {connect} from 'react-redux';
import {Link, Redirect} from 'react-router-dom';
import PropTypes from 'prop-types';

import OfferCard from '../offer-card/offer-card.jsx';

import ActionCreator from '../../action-creator/action-creator';

import {getUniqueCitiesNames} from '../../utils/utils';

const Favorites = ({currentCity, favorites, onCityClick}) => {
  const citiesNames = getUniqueCitiesNames(favorites);

  if (favorites.length === 0) {
    return <Redirect to='/favorites-empty' />;
  }

  return (
    <div className="page__favorites-container container">
      <section className="favorites">
        <h1 className="favorites__title">Saved listing</h1>
        <ul className="favorites__list">
          {citiesNames.map((city, i) => {
            return (
              <li className="favorites__locations-items" key={city + i + `favorites`}>
                <div className="favorites__locations locations locations--current">
                  <div className="locations__item">
                    <Link to={`/${city}`}
                      className="locations__item-link"
                      onClick={() => onCityClick(city)}>
                      <span>{city}</span>
                    </Link>
                  </div>
                </div>
                <div className="favorites__places">
                  {favorites
                    .filter((favoriteCity) => favoriteCity.city.name === city)
                    .map((el) => {
                      return (
                        <OfferCard
                          key={el.title + el.id}
                          cardInfo={el}
                        />
                      );
                    })}
                </div>
              </li>
            );
          })}
        </ul>
      </section>
      <footer className="footer container">
        <Link to={`/${currentCity}`} className="footer__logo-link">
          <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33" />
        </Link>
      </footer>
    </div>
  );
};

Favorites.propTypes = {
  currentCity: PropTypes.string,
  favorites: PropTypes.arrayOf(PropTypes.object).isRequired,
  onCityClick: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  currentCity: state.userState.city,
  favorites: state.data.favorites
});

const mapDispatchToProps = (dispatch) => ({
  onCityClick: (city) => {
    dispatch(ActionCreator.changeCity(city));
    dispatch(ActionCreator.changeSorting(`Popular`));
  },
});

export {Favorites};

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
