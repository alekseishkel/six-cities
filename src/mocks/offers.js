const offersCardsInfo = [{
  description: `Beautiful & luxurious apartment at great location`,
  img: `img/apartment-01.jpg`,
  price: `€120`,
  type: `Apartment`,
  isPremium: true
},
{
  description: `Wood and stone place`,
  img: `img/room.jpg`,
  price: `€80`,
  type: `Private room`,
  isPremium: false
},
{
  description: `Canal View Prinsengracht`,
  img: `img/apartment-02.jpg`,
  price: `€132`,
  type: `Apartment`,
  isPremium: false
},
{
  description: `Nice, cozy, warm big bed apartment`,
  img: `img/apartment-03.jpg`,
  price: `€180`,
  type: `Apartment`,
  isPremium: false
}];

export default offersCardsInfo;
