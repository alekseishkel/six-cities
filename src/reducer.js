const ActionCreator = {
  changeCity: (city) => ({
    type: `CHANGE_CITY`,
    payload: city
  }),

  changeCityCoords: (cityCoords) => ({
    type: `CHANGE_CITY_COORDS`,
    payload: cityCoords
  }),

  changeZoom: (zoom) => ({
    type: `CHANGE_ZOOM`,
    payload: zoom
  })
};

const initialState = {
  city: `Paris`,
  cityCoords: [48.8680266086781, 2.351173200195321],
  mapZoom: 11,
  offersCoords: [
    [52.3909553943508, 4.85309666406198],
    [52.369553943508, 4.85309666406198],
    [52.3909553943508, 4.929309666406198],
    [52.3809553943508, 4.939309666406198]
  ],
  offers: [{
    description: `Beautiful & luxurious apartment at great location`,
    img: `img/apartment-01.jpg`,
    price: `€120`,
    type: `Apartment`,
    isPremium: true,
    isBookmark: false
  },
  {
    description: `Wood and stone place`,
    img: `img/room.jpg`,
    price: `€80`,
    type: `Private room`,
    isPremium: false,
    isBookmark: false
  },
  {
    description: `Canal View Prinsengracht`,
    img: `img/apartment-02.jpg`,
    price: `€132`,
    type: `Apartment`,
    isPremium: false,
    isBookmark: false
  },
  {
    description: `Nice, cozy, warm big bed apartment`,
    img: `img/apartment-03.jpg`,
    price: `€180`,
    type: `Apartment`,
    isPremium: false,
    isBookmark: false
  }]
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case `CHANGE_CITY`:
      return Object.assign({}, state, {
        city: action.payload
      });
    case `CHANGE_CITY_COORDS`:
      return Object.assign({}, state, {
        cityCoords: action.payload
      });
    case `CHANGE_ZOOM`:
      return Object.assign({}, state, {
        mapZoom: action.payload
      });
    default:
      return state;
  }
};

export {
  ActionCreator,
  reducer
};
