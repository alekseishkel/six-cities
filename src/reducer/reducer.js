const ActionCreator = {
  changeCity: (city) => ({
    type: `CHANGE_CITY`,
    payload: city
  }),

  changeCityCoords: (cityCoords) => ({
    type: `CHANGE_CITY_COORDS`,
    payload: cityCoords
  }),

  changeZoom: (zoom) => ({
    type: `CHANGE_ZOOM`,
    payload: zoom
  })
};

const initialState = {
  city: `Paris`,
  cityCoords: [48.8680266086781, 2.351173200195321],
  mapZoom: 11,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case `CHANGE_CITY`:
      return Object.assign({}, state, {
        city: action.payload
      });
    case `CHANGE_CITY_COORDS`:
      return Object.assign({}, state, {
        cityCoords: action.payload
      });
    case `CHANGE_ZOOM`:
      return Object.assign({}, state, {
        mapZoom: action.payload
      });
    default:
      return state;
  }
};

export {
  ActionCreator,
  reducer
};
