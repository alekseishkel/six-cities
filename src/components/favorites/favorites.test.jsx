import React from 'react';
import renderer from 'react-test-renderer';

import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import reducer from '../../reducer/reducer';

import {Favorites} from './favorites.jsx';

import offers from '../../mocks/offers';

it(`Favorites correctly renders after relaunch`, () => {
  const store = createStore(reducer);

  const tree = renderer.create(
      <Provider store={store} >
        <Router>
          <Favorites
            currentCity={`Cologne`}
            favorites={[offers[0], offers[2], offers[4]]}
            onCityClick={jest.fn()}
          />
        </Router>
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
