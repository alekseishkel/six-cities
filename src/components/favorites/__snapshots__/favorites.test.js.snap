// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites correctly renders after relaunch 1`] = `
<div
  className="page__favorites-container container"
>
  <section
    className="favorites"
  >
    <h1
      className="favorites__title"
    >
      Saved listing
    </h1>
    <ul
      className="favorites__list"
    >
      <li
        className="favorites__locations-items"
      >
        <div
          className="favorites__locations locations locations--current"
        >
          <div
            className="locations__item"
          >
            <a
              className="locations__item-link"
              href="/Hamburg"
              onClick={[Function]}
            >
              <span>
                Hamburg
              </span>
            </a>
          </div>
        </div>
        <div
          className="favorites__places"
        >
          <article
            className="cities__place-card place-card"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <a
              href="/Hamburg/1"
              onClick={[Function]}
            >
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height={200}
                  src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/7.jpg"
                  width={260}
                />
              </div>
            </a>
            <div
              className="nullplace-card__info"
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    482
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button place-card__bookmark-buttonnull button"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height={19}
                    width={18}
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    In bookmarks
                  </span>
                </button>
              </div>
              <div
                className="place-card__rating rating"
              >
                <div
                  className="place-card__stars rating__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "55.99999999999999%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                The Pondhouse - A Magical Place
              </h2>
              <p
                className="place-card__type"
              >
                hotel
              </p>
            </div>
          </article>
          <article
            className="cities__place-card place-card"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <a
              href="/Hamburg/3"
              onClick={[Function]}
            >
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height={200}
                  src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/11.jpg"
                  width={260}
                />
              </div>
            </a>
            <div
              className="nullplace-card__info"
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    156
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button place-card__bookmark-buttonnull button"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height={19}
                    width={18}
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    In bookmarks
                  </span>
                </button>
              </div>
              <div
                className="place-card__rating rating"
              >
                <div
                  className="place-card__stars rating__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "88.00000000000001%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                Amazing and Extremely Central Flat
              </h2>
              <p
                className="place-card__type"
              >
                house
              </p>
            </div>
          </article>
        </div>
      </li>
      <li
        className="favorites__locations-items"
      >
        <div
          className="favorites__locations locations locations--current"
        >
          <div
            className="locations__item"
          >
            <a
              className="locations__item-link"
              href="/Cologne"
              onClick={[Function]}
            >
              <span>
                Cologne
              </span>
            </a>
          </div>
        </div>
        <div
          className="favorites__places"
        >
          <article
            className="cities__place-card place-card"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              className="place-card__mark"
            >
              <span>
                Premium
              </span>
            </div>
            <a
              href="/Cologne/5"
              onClick={[Function]}
            >
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height={200}
                  src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg"
                  width={260}
                />
              </div>
            </a>
            <div
              className="nullplace-card__info"
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                    232
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button place-card__bookmark-buttonnull button"
                  onClick={[Function]}
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height={19}
                    width={18}
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    In bookmarks
                  </span>
                </button>
              </div>
              <div
                className="place-card__rating rating"
              >
                <div
                  className="place-card__stars rating__stars"
                >
                  <span
                    style={
                      Object {
                        "width": "68%",
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                Wood and stone place
              </h2>
              <p
                className="place-card__type"
              >
                apartment
              </p>
            </div>
          </article>
        </div>
      </li>
    </ul>
  </section>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/Cologne"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
